src/mpid/ch4/src/ch4_coll.h

```diff
@@ -579,19 +586,23 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_Allgather_allcomm_composition_json(const void
 
     switch (cnt->id) {
         case MPIDI_CSEL_CONTAINER_TYPE__COMPOSITION__MPIDI_Allgather_intra_composition_alpha:
-            MPIDI_Allgather_fill_multi_leads_info(comm);
             /* make sure that the algo can be run */
-            if (MPIDI_COMM_ALLGATHER(comm, use_multi_leads) == 1) {
-                mpi_errno =
-                    MPIDI_Allgather_intra_composition_alpha(sendbuf, sendcount, sendtype,
-                                                            recvbuf, recvcount, recvtype,
-                                                            comm, errflag);
-            } else
-                mpi_errno =
-                    MPIR_Allgather_impl(sendbuf, sendcount, sendtype, recvbuf, recvcount, recvtype,
-                                        comm, errflag);
+            if (comm->comm_kind == MPIR_COMM_KIND__INTRACOMM)
+                MPIDI_Allgather_fill_multi_leads_info(comm);
+            MPII_COLLECTIVE_FALLBACK_CHECK(comm->rank,
+                                           MPIDI_COMM_ALLGATHER(comm, use_multi_leads) == 1 &&
+                                           (MPL_MAX(sendcount, recvcount) * type_size <=
```

Comment:
The send_type may be different from recv_type, so I don't think it is correct to just take the MAX (although it is safe). Why not precalcuation `data_sz` instead of `type_sz`?

Reply:
Does this look right?
```
MPIR_Datatype_get_size_macro(sendtype, send_type_size);
MPIR_Datatype_get_size_macro(recvtype, recv_type_size);
data_size = sendcount * send_type_size + recvcount * recv_type_size;
```
and ensure that
```
data_size <= MPIR_CVAR_ALLGATHER_SHM_PER_RANK
```

Same change for alltoall.

Reply:
If it is `MPI_IN_PLACE`, `data_sz = recv_count * recv_type_sz`, otherwise, it's `send_count * send_type_sz`.

Reply:
Thanks. Made this change in the code path without tuning file as well.